src/porepy/utils/ui_and_logging.py

```diff
@@ -1,63 +1,112 @@
-"""This module contains functionality to make the progress bars work correctly with
-logging."""
+"""This module contains functionality for logging and tqdm progressbars."""
 
 from __future__ import annotations
 
 import logging
 from contextlib import contextmanager
-from typing import Iterator, Optional, Type
+from typing import Iterator
+
+
+class DummyTrange:
+    """Dummy class to replace ``tqdm`` when it is not installed or used."""
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    @classmethod
+    def write(cls, *args, **kwargs):
+        # _TqdmLoggingHandler.emit calls this method to write messages. However, this
+        # functionality does not appear to be needed in the dummy class. Logging works
+        # fine with an empty dummy class.
+        pass
+
+    def update(self, *args, **kwargs):
+        pass
+
+    def set_description_str(self, *args, **kwargs):
+        pass
+
+    def set_postfix_str(self, *args, **kwargs):
+        pass
+
+    def close(self):
+        pass
+
 
 # ``tqdm`` is not a dependency. Up to the user to install it.
 try:
-    # Avoid some mypy trouble.
+    # Avoid some mypy trouble due to missing library stubs for tqdm.
     from tqdm.autonotebook import tqdm as std_tqdm  # type: ignore
-    from tqdm.contrib.logging import _TqdmLoggingHandler  # type: ignore
-    from tqdm.contrib.logging import logging_redirect_tqdm  # type: ignore
     from tqdm.contrib.logging import (  # type: ignore
         _get_first_found_console_logging_handler,
+        _TqdmLoggingHandler,
+        logging_redirect_tqdm,
     )
 except ImportError:
-    std_tqdm = None  # type: ignore
+    std_tqdm = DummyTrange
```

Comment:
Is this variable `std_tqdm` used anywhere else except for three lines below? I would suggest to either:
- use it to avoid the try/except blocks in the imports of the `nonlinear_solvers.py` and `run_models.py`. If so, I would rename it into something like `progressbar`.
- remove it and remove the default argument of `logging_redirect_tqdm_with_level`. 

The first option will make the mentioned files appear more clean, while the latter will be more explicit. Don't know what's best here.

Reply:
I like the first suggestion. Thanks!